lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -252,23 +300,41 @@ def find_matches(self, requirements):
                         raise ValueError(version_err) from ex
 
             return [
-                Candidate(fqcn, version, _none_src_server, first_req.type)
+                Candidate(fqcn, version, _none_src_server, first_req.type, None)
                 for version, _none_src_server in coll_versions
             ]
 
-        latest_matches = sorted(
-            {
-                candidate for candidate in (
-                    Candidate(fqcn, version, src_server, 'galaxy')
-                    for version, src_server in coll_versions
-                )
-                if all(self.is_satisfied_by(requirement, candidate) for requirement in requirements)
+        latest_matches = []
+        signatures = []
+        extra_signature_sources = []
+        for version, src_server in coll_versions:
+            tmp_candidate = Candidate(fqcn, version, src_server, 'galaxy', None)
+
+            unsatisfied = False
+            for requirement in requirements:
+                unsatisfied |= not self.is_satisfied_by(requirement, tmp_candidate)
                 # FIXME
-                # if all(self.is_satisfied_by(requirement, candidate) and (
-                #     requirement.src is None or  # if this is true for some candidates but not all it will break key param - Nonetype can't be compared to str
-                #     requirement.src == candidate.src
-                # ))
-            },
+                # unsatisfied |= not self.is_satisfied_by(requirement, tmp_candidate) or not (
+                #    requirement.src is None or  # if this is true for some candidates but not all it will break key param - Nonetype can't be compared to str
+                #    or requirement.src == candidate.src
+                # )
+                if unsatisfied:
+                    break
+                if not self._include_signatures:
+                    continue
+
+                extra_signature_sources.extend(requirement.signature_sources or [])
+
+            if not unsatisfied:
+                if self._include_signatures:
+                    signatures = src_server.get_collection_signatures(first_req.namespace, first_req.name, version)
+                    for extra_source in extra_signature_sources:
+                        signatures.append(self._artifacts_manager.get_signature_from_source(extra_source))
```

Comment:
As per the comment above, I agree that get_signature_from_source doesn't belong in the artifacts manager. I don't think it makes sense in the Galaxy API proxy either though, since these signatures have no relation to the galaxy server.

> Besides, if that idea from above to discover all of the signatures on construction time is implemented, you could maybe add an extra argument like append_from_disk or something like that, then part of this complexity could go away.

I responded to this above, but I can't think of a good way to do that.